core/services/relay/evm/read/multieventtype.go

```diff
@@ -0,0 +1,156 @@
+						ReadName:        eq.EventBinding.eventName,
+						ReturnVal:       eq.SequenceDataType,
+					})
+				}
+
+				err = newErrorFromCalls(err, calls, "", eventReadType)
+			} else {
+				err = fmt.Errorf("no event queries provided: %w", err)
+			}
+		}
+	}()
+
+	for _, eq := range eventQueries {
+		if err = eq.EventBinding.validateBound(eq.Address); err != nil {
+			return nil, err
+		}
+	}
+
+	allFilterExpressions := make([]query.Expression, 0, len(eventQueries))
+	for _, eq := range eventQueries {
+		var expressions []query.Expression
+
+		defaultExpressions := []query.Expression{
+			logpoller.NewAddressFilter(eq.Address),
+			logpoller.NewEventSigFilter(eq.EventBinding.hash),
+		}
+		expressions = append(expressions, defaultExpressions...)
+
+		remapped, remapErr := eq.EventBinding.remap(eq.Filter)
+		if remapErr != nil {
+			return nil, fmt.Errorf("error remapping filter: %w", err)
+		}
+		expressions = append(expressions, remapped.Expressions...)
+
+		filterExpression := query.And(expressions...)
+
+		allFilterExpressions = append(allFilterExpressions, filterExpression)
+	}
+
+	eventQuery := query.Or(allFilterExpressions...)
+
+	queryName := ""
+	for _, eq := range eventQueries {
+		queryName += eq.EventBinding.contractName + "-" + eq.Address.String() + "-" + eq.EventBinding.eventName + "-"
+	}
+	queryName = strings.TrimSuffix(queryName, "-")
+
+	logs, err := lp.FilteredLogs(ctx, []query.Expression{eventQuery}, limitAndSort, queryName)
+	if err != nil {
+		return nil, wrapInternalErr(err)
+	}
+
+	return decodeMultiEventTypeLogsIntoSequences(ctx, logs, eventQueries)
+}
+
+func decodeMultiEventTypeLogsIntoSequences(ctx context.Context, logs []logpoller.Log, eventQueries []EventQuery) (iter.Seq2[string, commontypes.Sequence], error) {
+	type sequenceWithKey struct {
+		Key      string
+		Sequence commontypes.Sequence
+	}
+	sequenceWithKeys := make([]sequenceWithKey, 0, len(logs))
+	eventSigToEventQuery := map[common.Hash]EventQuery{}
+	for _, eq := range eventQueries {
+		eventSigToEventQuery[eq.EventBinding.hash] = eq
+	}
+
+	for _, logEntry := range logs {
+		eventSignatureHash := logEntry.EventSig
+
+		eq, exists := eventSigToEventQuery[eventSignatureHash]
+		if !exists {
+			return nil, fmt.Errorf("no event query found for log with event signature %s", eventSignatureHash)
+		}
+
+		seqWithKey := sequenceWithKey{
+			Key: eq.Filter.Key,
+			Sequence: commontypes.Sequence{
+				Cursor: logpoller.FormatContractReaderCursor(logEntry),
+				Head: commontypes.Head{
+					Height:    strconv.FormatInt(logEntry.BlockNumber, 10),
+					Hash:      logEntry.BlockHash.Bytes(),
+					Timestamp: uint64(logEntry.BlockTimestamp.Unix()), //nolint:gosec // G115 false positive
+				},
+			},
+		}
+
+		var typeVal reflect.Value
+
+		typeInto := reflect.TypeOf(eq.SequenceDataType)
+		if typeInto.Kind() == reflect.Pointer {
+			typeVal = reflect.New(typeInto.Elem())
+		} else {
+			typeVal = reflect.Indirect(reflect.New(typeInto))
+		}
```

Comment:
would this panic if SequenceDataType is nil?

Reply:
it would, there are a couple of other places where the code could defensively verify input params to protect against nil panics, I'll update

Reply:
query parameters arg is now validate up front